<h1>Pre-procesamiento y limpieza de Datos</h1>

In [20]:
'''Importar la libreria para la realizacion de este laboratorio.'''
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 

In [48]:
'''Obtener el Dataset para mostrar todo su contenido.'''
df = pd.read_csv("data_act_01.csv",sep=";")
df.sample(10)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
6006  160942533                   Rtl  2016-04-03T00:00:00    16:33   
629   160920906                 Drugs  2016-04-01T00:00:00     9:04   
3588  160932246    Suspicious Vehicle  2016-04-02T00:00:00    15:34   
5653  160941898    Mentally Disturbed  2016-04-03T00:00:00    13:27   
9837  160963294         Audible Alarm  2016-04-05T00:00:00    18:33   
1022  160921701     Suspicious Person  2016-04-01T00:00:00    13:04   
1234  160922162           Petty Theft  2016-04-01T00:00:00    15:08   
8146  160952500                7.2.46  2016-04-04T00:00:00    15:24   
2087  160923716    Suspicious Vehicle  2016-04-01T00:00:00    22:13   
5941  160942409            Encampment  2016-04-03T00:00:00    15:55   

             CallDateTime Disposition                     Address  \
6006  2016-04-03T16:33:00         CAN         100 Block Of 6th St   
629   2016-04-01T09:04:00         HAN        100 Block Of Hyde St   
3588  2016-04-02T15:34:00         GOA        Portola Dr/fowler Av   
5653  2016-04-03T13:27:00         HAN            Irving St/7th Av   
9837  2016-04-05T18:33:00         PAS  300 Block Of Montgomery St   
1022  2016-04-01T13:04:00         ADV     200 Block Of Steiner St   
1234  2016-04-01T15:08:00         REP         Market St/church St   
8146  2016-04-04T15:24:00         NOM       2400 Block Of 17th Av   
2087  2016-04-01T22:13:00         GOA     500 Block Of Precita Av   
5941  2016-04-03T15:55:00         UTL        1400 Blk Harrison St   

               City State AgencyId  Range      AddressType  
6006  San Francisco    CA        1    NaN  Premise Address  
629   San Francisco    CA        1    NaN  Premise Address  
3588  San Francisco    CA        1    NaN     Intersection  
5653  San Francisco    CA        1    NaN     Intersection  
9837  San Francisco    CA        1    NaN  Premise Address  
1022  San Francisco    CA        1    NaN  Premise Address  
1234  San Francisco    CA        1    NaN     Intersection  
8146  San Francisco    CA        1    NaN  Premise Address  
2087  San Francisco    CA        1    NaN  Common Location  
5941  San Francisco    CA        1    NaN     Geo-Override

In [49]:
'''Observar el tipo de variables que contiene el dataframe'''
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10051 entries, 0 to 10050
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   CrimeId                10051 non-null  int64  
 1   OriginalCrimeTypeName  10051 non-null  object 
 2   OffenseDate            10051 non-null  object 
 3   CallTime               10051 non-null  object 
 4   CallDateTime           10051 non-null  object 
 5   Disposition            10051 non-null  object 
 6   Address                10051 non-null  object 
 7   City                   9730 non-null   object 
 8   State                  10048 non-null  object 
 9   AgencyId               10051 non-null  object 
 10  Range                  0 non-null      float64
 11  AddressType            10051 non-null  object 
dtypes: float64(1), int64(1), object(10)
memory usage: 942.4+ KB


In [50]:
'''Detectar valores nulos y su conteo'''
df.isnull().sum()

CrimeId                      0
OriginalCrimeTypeName        0
OffenseDate                  0
CallTime                     0
CallDateTime                 0
Disposition                  0
Address                      0
City                       321
State                        3
AgencyId                     0
Range                    10051
AddressType                  0
dtype: int64

In [51]:
'''Observar la cantidad de filas en el dataset.'''
df.shape

(10051, 12)

# Primer error (Columna Range)

In [52]:
'''Eliminar la columna "Range" ya que observando los registros se observa que no posee ningun dato.'''
df = df.drop(['Range'], axis=1)
df.sample(5)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
7395  160951092          Traffic Stop  2016-04-04T00:00:00     9:58   
9883  160963484        R.o. Violation  2016-04-05T00:00:00    19:22   
718   160921078     Suspicious Person  2016-04-01T00:00:00     9:51   
2454  160930225     Assault / Battery  2016-04-02T00:00:00     1:29   
5189  160941102     Suspicious Person  2016-04-03T00:00:00     9:30   

             CallDateTime Disposition                      Address  \
7395  2016-04-04T09:58:00         CIT       Stockton St/pacific Av   
9883  2016-04-05T19:22:00         GOA        3000 Block Of 23rd St   
718   2016-04-01T09:51:00         GOA     0 Block Of Montgomery St   
2454  2016-04-02T01:29:00          ND       2000 Block Of Union St   
5189  2016-04-03T09:30:00         CIT  200 Block Of Golden Gate Av   

               City State AgencyId      AddressType  
7395  San Francisco    CA        1     Intersection  
9883  San Francisco    CA        1  Premise Address  
718   San Francisco    CA        1  Premise Address  
2454  San Francisco    CA        1  Common Location  
5189  San Francisco    CA        1  Premise Address

# Segundo error (IDs duplicados)

In [53]:
'''Conteo de ids duplicados, y verificacion de los mismos.'''
print("CrimeId: ",df['CrimeId'].duplicated().sum())
dup = df[df['CrimeId'].duplicated(keep=False)]['CrimeId'].tolist()
dup

CrimeId:  4


[160913455, 160913455, 160913455, 160950496, 160950496, 160950496]

# Tercer error (Rellenar ciudades por valores)

In [54]:
'''Se reemplaza los valores vacios que existen 
en las columnas City por valores en Cero, 
ya que al tener valores nulos pueden generar
errores adicionales en este contexto.'''

df[['City']] = df[['City']].fillna(value=0)
df

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1      160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2      160912590             Susp Info  2016-03-31T00:00:00    16:49   
3      160912801                Report  2016-03-31T00:00:00    17:38   
4      160912811                   594  2016-03-31T00:00:00    17:42   
...          ...                   ...                  ...      ...   
10046  160964210      Well Being Check  2016-04-05T00:00:00    23:38   
10047  160964216          Traffic Stop  2016-04-05T00:00:00    23:43   
10048  160964227          Traffic Stop  2016-04-05T00:00:00    23:48   
10049  160964229          Passing Call  2016-04-05T00:00:00    23:48   
10050  160964249          Traffic Stop  2016-04-05T00:00:00    23:54   

              CallDateTime   Disposition                  Address  \
0      2016-03-30T18:42:00           REP  100 Block Of Chilton Av   
1      2016-03-31T15:31:00           GOA  2300 Block Of Market St   
2      2016-03-31T16:49:00           GOA  2300 Block Of Market St   
3      2016-03-31T17:38:00           GOA      500 Block Of 7th St   
4      2016-03-31T17:42:00           REP       Beale St/bryant St   
...                    ...           ...                      ...   
10046  2016-04-05T23:38:00           GOA    Hollister Av/hawes St   
10047  2016-04-05T23:43:00           ADV   California St/mason St   
10048  2016-04-05T23:48:00           CIT   Rosella Ct/onondaga Av   
10049  2016-04-05T23:48:00  Not recorded          0 Blk Herbst Rd   
10050  2016-04-05T23:54:00           ADV        Grant Av/geary St   

                City State AgencyId      AddressType  
0      San Francisco    CA        1  Premise Address  
1      San Francisco    CA        1  Premise Address  
2      San Francisco    CA        1  Premise Address  
3      San Francisco    CA        1  Premise Address  
4      San Francisco    CA        1     Intersection  
...              ...   ...      ...              ...  
10046  San Francisco    CA        1     Intersection  
10047  San Francisco    CA        1     Intersection  
10048  San Francisco    CA        1     Intersection  
10049  San Francisco    CA        1     Geo-Override  
10050  San Francisco    CA        1     Intersection  

[10051 rows x 11 columns]

In [55]:
'''Verificar la redundancia de datos en la columna de City'''
df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
0                  321                    321          321       321   
 S                   1                      1            1         1   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
SAN FRANCISCO        1                      1            1         1   
San Francisco     9665                   9665         9665      9665   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               CallDateTime  Disposition  Address  State  AgencyId  \
City                                                                 
0                       321          321      321    319       321   
 S                        1            1        1      0         1   
Brisbane                  1            1        1      1         1   
Daly City                 5            5        5      5         5   
Presidio                  3            3        3      3         3   
SAN FRANCISCO             1            1        1      1         1   
San Francisco          9665         9665     9665   9665      9665   
Treasure Isla            51           51       51     51        51   
Yerba Buena               3            3        3      3         3   

               AddressType  
City                        
0                      321  
 S                       1  
Brisbane                 1  
Daly City                5  
Presidio                 3  
SAN FRANCISCO            1  
San Francisco         9665  
Treasure Isla           51  
Yerba Buena              3

<h1>Analisis</h1>
Observamos que existian 321 valores nulos que fueron reemplazados por 0 Analizando las direccciones de estas filas, nos podemos dar cuenta que todas pertenecen a la ciudad de San Francisco, asi que reemplazaremos estos Valores por San Francisco. Del mismo modo se observan dos errores, existe una columna con la letra S, probablemente de San Francisco, comprobaremos la direccion de esta para estar seguros de que la ciudad es San Francisco. Asi mismo observamos que existe un dato con mayusculas. Vamos a corregir ese error.

In [56]:
'''Se reemplaza el dato en mayusculas, por minusculas.'''
df['City'] = df['City'].replace('SAN FRANCISCO','San Francisco');

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1      160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2      160912590             Susp Info  2016-03-31T00:00:00    16:49   
3      160912801                Report  2016-03-31T00:00:00    17:38   
4      160912811                   594  2016-03-31T00:00:00    17:42   
...          ...                   ...                  ...      ...   
10046  160964210      Well Being Check  2016-04-05T00:00:00    23:38   
10047  160964216          Traffic Stop  2016-04-05T00:00:00    23:43   
10048  160964227          Traffic Stop  2016-04-05T00:00:00    23:48   
10049  160964229          Passing Call  2016-04-05T00:00:00    23:48   
10050  160964249          Traffic Stop  2016-04-05T00:00:00    23:54   

              CallDateTime   Disposition                  Address  \
0      2016-03-30T18:42:00           REP  100 Block Of Chilton Av   
1      2016-03-31T15:31:00           GOA  2300 Block Of Market St   
2      2016-03-31T16:49:00           GOA  2300 Block Of Market St   
3      2016-03-31T17:38:00           GOA      500 Block Of 7th St   
4      2016-03-31T17:42:00           REP       Beale St/bryant St   
...                    ...           ...                      ...   
10046  2016-04-05T23:38:00           GOA    Hollister Av/hawes St   
10047  2016-04-05T23:43:00           ADV   California St/mason St   
10048  2016-04-05T23:48:00           CIT   Rosella Ct/onondaga Av   
10049  2016-04-05T23:48:00  Not recorded          0 Blk Herbst Rd   
10050  2016-04-05T23:54:00           ADV        Grant Av/geary St   

                City State AgencyId      AddressType  
0      San Francisco    CA        1  Premise Address  
1      San Francisco    CA        1  Premise Address  
2      San Francisco    CA        1  Premise Address  
3      San Francisco    CA        1  Premise Address  
4      San Francisco    CA        1     Intersection  
...              ...   ...      ...              ...  
10046  San Francisco    CA        1     Intersection  
10047  San Francisco    CA        1     Intersection  
10048  San Francisco    CA        1     Intersection  
10049  San Francisco    CA        1     Geo-Override  
10050  San Francisco    CA        1     Intersection  

[10051 rows x 11 columns]

In [58]:
'''Se reemplaza ahora la letra S por la ciudad de San Francisco'''
df['City']= df['City'].replace('S','San Francisco')

In [60]:
'''Se reemplaza ahora las filas de 0 por la ciudad de San Francisco'''
df['City']= df['City'].replace(0,'San Francisco')

In [61]:
'''Verificar la redundancia de datos en la columna de City otra vez'''
df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
San Francisco     9988                   9988         9988      9988   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               CallDateTime  Disposition  Address  State  AgencyId  \
City                                                                 
Brisbane                  1            1        1      1         1   
Daly City                 5            5        5      5         5   
Presidio                  3            3        3      3         3   
San Francisco          9988         9988     9988   9985      9988   
Treasure Isla            51           51       51     51        51   
Yerba Buena               3            3        3      3         3   

               AddressType  
City                        
Brisbane                 1  
Daly City                5  
Presidio                 3  
San Francisco         9988  
Treasure Isla           51  
Yerba Buena              3

In [62]:
'''Verificamos nuevamente los valores nulos, 
como se observa existen 3 estados que son nulos.'''
df.isnull().sum()

CrimeId                  0
OriginalCrimeTypeName    0
OffenseDate              0
CallTime                 0
CallDateTime             0
Disposition              0
Address                  0
City                     0
State                    3
AgencyId                 0
AddressType              0
dtype: int64